CONTRACTS: loop assigns clause inference function

remi-delmas-3000 opened the pull request:
The first commit is from #7636, do not review.

Co-authored with @qinheping

Loop assigns clause inference function. Uses `loop_utils.h` to get a set of lvalue expressions modulo aliasing that may be assigned by the loop. 
The lvalue expressions are then post processed. 

- If an inferred target depends on a loop-local variable (i.e. a variable that does not exist outside of the loop body) it is widened to its base object. If the base object still depends on a loop local, the target is discarded, otherwise `__CPROVER_object_whole(&root_object)` is kept as inferred target.
- If the address of an inferred target is non-constant, then `__CPROVER_object_whole(&target)` is used as widened target.

Extracted from https://github.com/diffblue/cbmc/pull/7541, can only be tested once all features are in place.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Commit:
CONTRACTS: loop assigns clause inference function

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h:
```diff
@@ -0,0 +1,34 @@
+
+// \brief Infer assigns clause targets for a loop from its instructions and a
+// may alias analysis at the function level.
+std::set<exprt> dfcc_infer_loop_assigns(
```
Is a `set` the most appropriate data structure here? Do you actually need the ordering? I'd otherwise suggest to use `std::unordered_set`.

remi-delmas-3000 replied:
The order is not relevant, but this is due to `loop_util.h` returning `assignst` defined as `typedef std::set<exprt> assignst`. 

I'd rather change the return type to `assignst` instead of adding a translation from `set` to `unordered_set`. Would that be OK ?

tautschnig replied:
Shouldn't this now be `assignst` as you suggest?

remi-delmas-3000 replied:
fixed.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h:
```diff
@@ -0,0 +1,34 @@
+  const loopt &loop_instructions,
+  const source_locationt &loop_head_location,
+  messaget &log,
+  namespacet &ns);
```
Should be `const`

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+  try
+  {
+    // infer
+    assignst assigns;
+    get_assigns(local_may_alias, loop_instructions, assigns);
```
Can we change `get_assigns` so that it returns by value?

remi-delmas-3000 replied:
`loop_utils.h` defines `get_assigns` like this, and the whole k-induction module depends on this interface so I'd rather not change it.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+    get_assigns(local_may_alias, loop_instructions, assigns);
+
+    // compute locals
+    std::set<irep_idt> loop_locals;
```
Probably `std::unordered_set`

remi-delmas-3000 replied:
changed to `assignst`

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+    auto expr_it = assigns.begin();
+    while(expr_it != assigns.end())
+    {
+      const std::unordered_set<irep_idt> symbols =
+        find_symbol_identifiers(*expr_it);
```
Is it possible for `*expr_it` to be, e.g., `A[i]`? If so, wouldn't `i` wrongly end up in this set? Don't you actually want what is being built in #7626?

remi-delmas-3000 replied:
Yes if an expression is of the form `a[i]` with `i` non constant or a loop local, post processing is applied to widen the inferred target to the underlying root object. #7636 is now included as first commit in this PR.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+      const std::unordered_set<irep_idt> symbols =
+        find_symbol_identifiers(*expr_it);
+
+      if(std::find_if(symbols.begin(), symbols.end(), [&](const irep_idt &id) {
+           return loop_locals.find(id) != loop_locals.end();
+         }) != symbols.end())
+      {
+        expr_it = assigns.erase(expr_it);
+      }
```
Is it correct to remove an expression from `assigns` if _any_ symbol occurring in there is a loop-local one? (At least that is my understanding of what the `find_if` will accomplish here.)

remi-delmas-3000 replied:
Removing something from assigns is always sound as it can only cause the analysis to fail with an assigns clause violation.

We really have to remove any inferred expression that contains a non-local symbol, because assigns clause targets can only talk about variables that exist outside of the loop.

```
for (int i = 0; i < N; i++) {
 int j = i;
 arr[j] = 0;
}
```
In that case the inference returns the expressions `i` and `arr[j]` as inferred targets, and we cannot use `arr[j]`.

Maybe we could do the following instead: if a target depends on some loop local, compute `root = root_object(inferred_target)`,  if `root` still contains non-local variables, discard it, otherwise widen to `__CPROVER_object_whole(root)` and add it to the set.
 
  @qinheping does that sound right ?

qinheping replied:
In the example, `get_assigns` will give us `j` and `arr[j]`. 
Yes, we should remove the local variable `j`.
For deference expressions with non-constant offset, e.g., `*(arr+j)`; and indexing expressions with non-constant index, e.g., `arr[j]`, we should widen them with `__CPROVER_object_whole`.

remi-delmas-3000 replied:
This is what is implemented in this new revision, thanks to #7636

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+    dfcc_widen_non_const(ns, assigns);
+    return assigns;
+  }
+  catch(const analysis_exceptiont &exc)
```
Which of the above code can actually throw? It seems it could only be `get_assigns`, but I wasn't able to find a `throw` over there either.

qinheping replied:
I removed the `throw` in #7603 as even we failed to compute alias assigns, it is still meaningful to return what we infer.

remi-delmas-3000 replied:
so should we remove the try/catch completely ?

qinheping replied:
Yes, we should remove the try/catch.

remi-delmas-3000 replied:
I removed the try-catch

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+{
+  assignst result;
+  const symbolt &object_whole_sym = ns.lookup(CPROVER_PREFIX "object_whole");
+  code_typet object_whole_code_type = to_code_type(object_whole_sym.type);
```
`const code_typet &`

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+          object_whole_sym.symbol_expr(),
+          args,
+          object_whole_code_type.return_type(),
+          address_of_expr.source_location()));
```
No source location is set for `address_of_expr`, so this will always be `nil`.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+        exprt::operandst args;
+        args.emplace_back(address_of_expr);
+        result.emplace(side_effect_expr_function_callt(
+          object_whole_sym.symbol_expr(),
```
The same `symbol_expr` will be constructed repeatedly.

remi-delmas-3000 replied:
How can I prevent that ?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,119 @@
+      // index or offset is non-constant.
+      if(!is_constant(address_of_expr))
+      {
+        exprt::operandst args;
+        args.emplace_back(address_of_expr);
```
`exprt::operandst args{{address_of_expr}};` should facilitate RAII.

remi-delmas-3000 replied:
fixed

Commit:
CONTRACTS: functions to compute root objects

Alternative implementation of `root_object`
supporting ternary operators in assignment or call
LHS expressions, and object slice expressions in
assigns clause targets.

remi-delmas-3000 commented:
@tautschnig @qinheping I just pushed a revised version that uses root objects to widen targets, should be stable now

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h:
```diff
@@ -0,0 +1,46 @@
+class namespacet;
+class messaget;
+
+/// Computes a set of root object expressions for an assignment lhs expression.
```
```suggestion
/// Computes a set of root object expressions for an assignment `lhs` expression.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.h:
```diff
@@ -0,0 +1,46 @@
+  messaget &log);
+
+/// Computes a single root object expressions for an assigns clause target that
+/// is expected not to contain disjunctions. Fails if the exprssion contains
```
```suggestion
/// is expected not to contain disjunctions. Fails if the expression contains
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,100 @@
+    expr.source_location());
+}
+
+static bool
+depends_on(const exprt &expr, std::unordered_set<irep_idt> identifiers)
```
Could you also add a comment header here?

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp:
```diff
@@ -0,0 +1,190 @@
+
+/// Recursively computes a set of root object expressions for an assignment lhs.
+static void lhs_root_objects_rec(
+  const exprt &expr,
```
```suggestion
  const exprt &lhs,
```

remi-delmas-3000 replied:
fixed doc

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp:
```diff
@@ -0,0 +1,190 @@
+
+#include <langapi/language_util.h>
+
+/// Recursively computes a set of root object expressions for an assignment lhs.
```
```suggestion
/// Recursively computes a set of root object expressions for an assignment `lhs`.
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+std::unordered_set<exprt, irep_hash> dfcc_assignment_lhs_root_objects(
+  const exprt &expr,
```
```suggestion
  const exprt &lhs,
```

remi-delmas-3000 replied:
fixed

feliperodri commented on src/goto-instrument/contracts/dynamic-frames/dfcc_root_object.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+/// Translates object slice expressions found in assigns clause targets to
+/// dereference expressions, so that object_descriptor_exprt can be used to
```
```suggestion
/// dereference expressions, so that `object_descriptor_exprt` can be used to
```

remi-delmas-3000 replied:
fixed

qinheping reviewed (approved):
LGTM

feliperodri commented:
@tautschnig if you have the time, could you do another round of review on this PR?

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.h:
```diff
@@ -0,0 +1,30 @@
+  const local_may_aliast &local_may_alias,
+  const loopt &loop_instructions,
+  const source_locationt &loop_head_location,
+  messaget &log,
```
We usually use a `message_handlert`. But, also, you might get rid of it once the comments in the depended-on PR are addressed.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,103 @@
+      loop_locals.insert(target->decl_symbol().get_identifier());
+  }
+
+  // widen or drop targets that depend on loop-locals or are non-constant
```
I had to read the remainder of the code to understand what was meant by "non-constant" here. I think this needs a bit more text.

remi-delmas-3000 replied:
added example. But this notion is actually implemented in `havoc_utils_is_constantt`

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,103 @@
+      address_of_expr.add_source_location() = expr.source_location();
+      if(!is_constant(address_of_expr))
+      {
+        // Target address is not constant, widening to the whole object
+        result.emplace(make_object_whole_call_expr(address_of_expr, ns));
```
What happens behind the scenes with a pointer passed to object-whole? Would it be the same as the `root_object` case above?

remi-delmas-3000 replied:
The meaning of  `object_whole(&expr)` or `object_whole(&root_object(expr))` is the same so we don't bother computing the root object. In the first case we compute the root_object to try and get rid of a dependency on loop locals or drop the target if the root object is still local to the loop. In the second case we have a non-constant expression that does not depend on loop locals, and we just widen it directly.

remi-delmas-3000 replied:
we have to use `object_whole` because this is the predicate that DFCC understands to specify byte ranges.

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,103 @@
+  exprt::operandst args{{expr}};
+  return side_effect_expr_function_callt(
+    object_whole_sym.symbol_expr(),
+    args,
```
Just use `{{expr}}` or else `std::move(args)`.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/dynamic-frames/dfcc_infer_loop_assigns.cpp:
```diff
@@ -0,0 +1,103 @@
+#include <util/std_code.h>
+
+#include <goto-instrument/havoc_utils.h>
+#include <langapi/language_util.h>
```
Is this needed?

remi-delmas-3000 replied:
cleaned up

tautschnig reviewed (commented):
Just minor cleanup, will be good to go once the depended-on PR is merged.

remi-delmas-3000 commented:
@tautschnig all suggestions implemented, could you please take a last look ?
